File: packages/python/mythos_sdk/listing_callback.py
import logging
import os
from collections.abc import Awaitable, Callable

from fastapi import Request
from fastapi.responses import JSONResponse
from jose import jwt
from jose.exceptions import ExpiredSignatureError, JOSEError, JWTClaimsError, JWTError

from .jwks_cache import get_jwks, get_jwks_with_kid_fallback

_logger = logging.getLogger(__name__)

_DEFAULT_API_URL = "https://api.mythos.work"
_MYTHOS_ISSUER = "mythos"
_DECODE_OPTIONS = {"verify_aud": False}


async def _validate_listing_callback_token(token: str) -> str:
    api_url = os.environ.get("MYTHOS_API_URL", _DEFAULT_API_URL)
    jwks = await get_jwks(api_url)
    try:
        payload = jwt.decode(
            token, jwks, algorithms=["RS256"], issuer=_MYTHOS_ISSUER, options=_DECODE_OPTIONS
        )
    except ExpiredSignatureError:
        raise
    except JWTError:
        jwks = await get_jwks_with_kid_fallback(api_url)
        payload = jwt.decode(
            token, jwks, algorithms=["RS256"], issuer=_MYTHOS_ISSUER, options=_DECODE_OPTIONS
        )

    if payload.get("purpose") != "listing_registered":
        raise JWTClaimsError("Token purpose is not listing_registered")

    return payload["listingId"]


def create_listing_callback_handler(
    on_registered: Callable[[str], Awaitable[None]],
) -> Callable[[Request], Awaitable[JSONResponse]]:
    async def mythos_listing_registered(request: Request) -> JSONResponse:
        token = request.query_params.get("lt")
        if not token:
            return JSONResponse({"error": "Missing listing callback token"}, status_code=401)
        try:
            listing_id = await _validate_listing_callback_token(token)
            await on_registered(listing_id)
        except JOSEError:
            return JSONResponse({"error": "Invalid listing callback token"}, status_code=401)
        except Exception:
            _logger.exception("listing-callback: unexpected error")
            return JSONResponse({"error": "Service unavailable"}, status_code=503)
        return JSONResponse({"ok": True})

    return mythos_listing_registered

File: packages/node/src/listing-callback.ts
import { jwtVerify, errors } from 'jose';
import type { RequestHandler } from 'express';
import { getKeySet, getKeySetWithKidFallback } from './jwks-cache';

const MYTHOS_ISSUER = 'mythos';
const DEFAULT_API_URL = 'https://api.mythos.work';

async function validateListingCallbackToken(token: string): Promise<string> {
  const apiUrl = process.env.MYTHOS_API_URL ?? DEFAULT_API_URL;
  let keySet = await getKeySet(apiUrl);

  let payload;
  try {
    ({ payload } = await jwtVerify(token, keySet, { algorithms: ['RS256'], issuer: MYTHOS_ISSUER }));
  } catch (err: unknown) {
    if (!(err instanceof errors.JWKSNoMatchingKey)) throw err;
    keySet = await getKeySetWithKidFallback(apiUrl);
    ({ payload } = await jwtVerify(token, keySet, { algorithms: ['RS256'], issuer: MYTHOS_ISSUER }));
  }

  if (payload['purpose'] !== 'listing_registered') {
    throw new errors.JWTClaimValidationFailed(
      'Token purpose is not listing_registered',
      payload,
      'purpose',
      'check',
    );
  }

  return payload['listingId'] as string;
}

export function listingCallbackRoute(
  onRegistered: (listingId: string) => Promise<void>,
): RequestHandler {
  return async (req, res) => {
    const raw = req.query['lt'];
    const first = Array.isArray(raw) ? raw[0] : raw;
    const token = typeof first === 'string' ? first : undefined;
    if (!token) {
      res.status(401).json({ error: 'Missing listing callback token' });
      return;
    }
    try {
      const listingId = await validateListingCallbackToken(token);
      await onRegistered(listingId);
    } catch (err) {
      if (err instanceof errors.JOSEError) {
        res.status(401).json({ error: 'Invalid listing callback token' });
      } else {
        console.error('[mythos-sdk] listing-callback: unexpected error', err);
        res.status(503).json({ error: 'Service unavailable' });
      }
      return;
    }
    res.json({ ok: true });
  };
}
